pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyCRX.hpp

```diff
@@ -0,0 +1,289 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines CRX gate
+ */
+#pragma once
+#include "AVXUtil.hpp"
+#include "BitUtil.hpp"
+#include "Blender.hpp"
+#include "Permutation.hpp"
+
+#include "ConstantUtil.hpp"
+#include "Util.hpp"
+
+#include <immintrin.h>
+
+#include <complex>
+#include <utility>
+
+namespace Pennylane::Gates::AVXCommon {
+template <typename PrecisionT, size_t packed_size> struct ApplyCRX {
+    using Precision = PrecisionT;
+    using PrecisionAVXConcept = AVXConceptType<PrecisionT, packed_size>;
+
+    constexpr static auto packed_size_ = packed_size;
+    constexpr static bool symmetric = false;
+
+    template <size_t control, size_t target>
+    static constexpr auto applyInternalInternalPermutation() {
+        std::array<uint8_t, packed_size> perm{};
+
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) { // if control bit is 1
+                perm[2 * k + 0] = 2 * (k ^ (1U << target)) + 1;
+                perm[2 * k + 1] = 2 * (k ^ (1U << target)) + 0;
+            } else {
+                perm[2 * k + 0] = 2 * k + 0;
+                perm[2 * k + 1] = 2 * k + 1;
+            }
+        }
+        return Permutation::compilePermutation<PrecisionT>(perm);
+    }
+
+    template <size_t control, size_t target, class ParamT>
+    static auto applyInternalInternalOffDiagFactor(ParamT angle) {
+        std::array<PrecisionT, packed_size> arr{};
+
+        // positions are after permutations
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) { // if control bit is 1
+                arr[2 * k + 0] = std::sin(angle / 2);
+                arr[2 * k + 1] = -std::sin(angle / 2);
+            } else {
+                arr[2 * k + 0] = Precision{0.0};
+                arr[2 * k + 1] = Precision{0.0};
+            }
+        }
+        return set<Precision, packed_size>(arr);
+    }
+
+    template <size_t control, size_t target, class ParamT>
+    static auto applyInternalInternalDiagFactor(ParamT angle) {
```

Comment:
Could you please add docstring here? It would be even better if some explainations here on the difference among those 'applyInternalInternalDiagFactor', 'applyInternalInternalOffDiagFactor' and 'applyInternalExternal'.

Reply:
I agree with @multiphaseCFD. We'll need more docstrings here and in the other files.